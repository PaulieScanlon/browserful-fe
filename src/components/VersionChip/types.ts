export enum overrideTypes {
  IS_INCLUDED = 'isIncluded',
  IS_EXCLUDED = 'isExcluded'
}
