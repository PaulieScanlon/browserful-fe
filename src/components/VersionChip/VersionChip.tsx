import * as React from 'react';

import {
  VersionLabel,
  VersionText,
  VersionInput,
  VersionStyle
} from './styles';

interface IProps {
  browser: string;
  version: any;
  onChange: (
    event: React.ChangeEventHandler<HTMLInputElement>,
    browser: string,
    version: number | string
  ) => void;
  isIncluded?: boolean;
  defaultChecked?: boolean;
}

export const VersionChip: React.SFC<IProps> = ({
  version,
  browser,
  onChange,
  isIncluded,
  defaultChecked
}: IProps) => {
  return (
    <VersionLabel htmlFor={browser}>
      <VersionInput
        defaultChecked={defaultChecked}
        type="checkbox"
        id={browser}
        onChange={event => onChange(event as any, browser, version)}
        isIncluded={isIncluded}
      />
      <VersionStyle>
        <VersionText isIncluded={isIncluded}>{version || 0}</VersionText>
      </VersionStyle>
    </VersionLabel>
  );
};

VersionChip.defaultProps = {
  isIncluded: false,
  defaultChecked: false
};
