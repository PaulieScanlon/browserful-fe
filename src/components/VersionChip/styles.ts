import styled from 'react-emotion';

import { font } from '../../typography';
import { colours } from '../../theme';

interface IProps {
  isIncluded: boolean;
}

const getStyleRules = (isIncluded, checked) => {
  if (isIncluded) {
    return {
      borderColor: checked ? colours.green : colours.transparent,
      backgroundColor: colours.greenLight
    };
  } else {
    return {
      borderColor: checked ? colours.red : colours.transparent,
      backgroundColor: colours.redLight
    };
  }
};

export const VersionLabel = styled.label({
  label: 'version-label',
  position: 'relative',
  display: 'inline-flex',
  cursor: 'pointer',
  justifyContent: 'center',
  alignItems: 'center'
});

export const VersionText = styled.span<IProps>(
  {
    fontSize: '12px',
    lineHeight: '14px',
    position: 'absolute',
    fontFamily: `${font.fontFamily}`,
    textAlign: 'center',
    whiteSpace: 'pre-wrap',
    textTransform: 'capitalize'
  },
  ({ isIncluded }) => ({
    fontWeight: isIncluded ? 'bold' : 'normal',
    color: isIncluded ? colours.green : colours.red
  })
);

export const VersionInput = styled.input<IProps>(
  {
    label: 'verion-input',
    position: 'absolute',
    display: 'none'
  },
  ({ isIncluded }: IProps) => ({
    ':checked + span': {
      ...getStyleRules(isIncluded, true)
    },
    '+ span': {
      ...getStyleRules(isIncluded, false)
    }
  })
);

export const VersionStyle = styled.span({
  label: 'version-style',
  width: '40px',
  height: '40px',
  borderRadius: '20px',
  borderStyle: 'solid',
  boxSizing: 'border-box',
  outline: 'none'
});
