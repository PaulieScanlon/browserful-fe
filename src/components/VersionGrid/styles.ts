import styled from 'react-emotion';
import { spaceSm } from '../../theme';

export const GridContent = styled.div({
  label: 'gird-content',

  '> * ': {
    margin: `${spaceSm}px`
  }
});
