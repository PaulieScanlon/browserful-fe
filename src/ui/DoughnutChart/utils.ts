import { IData } from './DoughnutChart';

export const addFillToData = (data: Array<IData>) => {
  return data.map(data => {
    return {
      name: data.name,
      value: data.value,
      fill: data.fill,
      stroke: data.stroke
    };
  });
};
