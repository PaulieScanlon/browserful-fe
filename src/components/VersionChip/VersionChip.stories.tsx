import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';

import { VersionChip } from './VersionChip';

import browserslist from 'browserslist';
import { friendlyIfy } from '../../utils/friendlyIfy';

const mockData = friendlyIfy(browserslist(['last 1 Chrome versions']));

const stories = storiesOf('VersionChip', module);

const onChange = (event, browser, version) => {
  action('onChange')(event.currentTarget, browser, version);
};

stories.add(
  'default usage',
  withInfo('The browser and version prop are used on the onChange callback')(
    () => (
      <VersionChip
        browser={mockData[0].name}
        version={mockData[0].versions[0].id}
        onChange={(event, browser, version) =>
          onChange(event, browser, version)
        }
      />
    )
  )
);

stories.add(
  '!isIncluded & defaultChecked',
  withInfo('The defaultChecked prop is the checked status of the checkbox')(
    () => (
      <VersionChip
        defaultChecked={true}
        browser={mockData[0].name}
        version={mockData[0].versions[0].id}
        onChange={(event, browser, version) =>
          onChange(event, browser, version)
        }
      />
    )
  )
);

stories.add(
  'isIncluded',
  withInfo('The isIncluded prop shows if a browser/version is included')(() => (
    <VersionChip
      isIncluded={true}
      browser={mockData[0].name}
      version={mockData[0].versions[0].id}
      onChange={(event, browser, version) => onChange(event, browser, version)}
    />
  ))
);

stories.add(
  'isIncluded & defaultChecked',
  withInfo('The defaultChecked prop is the checked status of the checkbox')(
    () => (
      <VersionChip
        defaultChecked={true}
        isIncluded={true}
        browser={mockData[0].name}
        version={mockData[0].versions[0].id}
        onChange={(event, browser, version) =>
          onChange(event, browser, version)
        }
      />
    )
  )
);
