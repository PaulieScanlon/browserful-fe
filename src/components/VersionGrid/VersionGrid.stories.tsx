import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';

import browserslist from 'browserslist';
import { friendlyIfy } from '../../utils/friendlyIfy';

const mockData = friendlyIfy(
  browserslist(browserslist(['last 999 Chrome versions']))
);

import { VersionGrid } from './VersionGrid';

const stories = storiesOf('VersionGrid', module);

const onClick = (event, browser, version) => {
  action('onClick')(event.currentTarget, browser, version);
};

stories.add(
  'default usage',
  withInfo('VersionGrid displays versionChips and a other stuff..')(() => (
    <VersionGrid
      data={mockData[0]}
      onClick={(event, browser, version) => onClick(event, browser, version)}
    />
  ))
);
