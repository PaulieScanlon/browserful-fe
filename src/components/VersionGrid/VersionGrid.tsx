import * as React from 'react';

import { GridContent } from './styles';

import { VersionChip } from '../VersionChip';

interface IProps {
  data: any;
  maxHeight?: string;
  defaultChecked?: boolean;
  onClick?: (
    event: React.ChangeEventHandler<HTMLInputElement>,
    browser: string,
    version: number
  ) => void;
}

export const VersionGrid: React.SFC<IProps> = ({ data, onClick }: IProps) => {
  const { name, versions } = data;

  const versionChips = versions.map((obj, i) => {
    return (
      <VersionChip
        key={i}
        browser={`${name}-${obj.id}`}
        // @TODO, this string replace should probs happen in the util function
        version={obj.id.replace('-', '\n')}
        isIncluded={obj.isIncluded}
        defaultChecked={obj.defaultChecked}
        onChange={event => onClick(event, name, obj.id)}
      />
    );
  });

  return <GridContent>{versionChips}</GridContent>;
};

VersionGrid.defaultProps = {
  onClick: () => {}
};
