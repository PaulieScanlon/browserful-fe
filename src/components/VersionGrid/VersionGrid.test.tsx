import * as React from 'react';
import { shallow } from 'enzyme';

import browserslist from 'browserslist';
import { createMatrix } from '../../utils/createMatrix';

const mockData = createMatrix(browserslist(['last 999 Firefox versions']));

import { VersionGrid } from './VersionGrid';
import { VersionChip } from '../VersionChip';

const onClick = (event, browser, version) => {};

const shallowDefault = () => {
  return shallow(
    <VersionGrid
      data={mockData[0]}
      onClick={(event, browser, version) => onClick(event, browser, version)}
    />
  );
};

describe('<VersionGrid />', () => {
  it('renders correctly', () => {
    const wrapper = shallowDefault();
    expect(wrapper).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls version function on click', () => {
    const testOnOnClick = jest.fn();
    const wrapper = shallow(
      <VersionGrid
        data={mockData[0]}
        onClick={(event, browser, version) =>
          testOnOnClick(event, browser, version)
        }
      />
    );
    const versionChip = wrapper.find(VersionChip).at(0);
    versionChip.simulate('change');
    expect(testOnOnClick).toHaveBeenCalledTimes(1);
  });
});
