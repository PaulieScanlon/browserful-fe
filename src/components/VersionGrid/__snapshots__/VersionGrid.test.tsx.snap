// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionGrid /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VersionGrid
    data={
      Object {
        "defaultChecked": true,
        "friendlyName": "Chrome (Android)",
        "logo": "chrome-for-android",
        "name": "and_chr",
        "percent": undefined,
        "platform": "mobile",
        "versions": Array [
          Object {
            "defaultChecked": false,
            "id": "69",
            "isIncluded": false,
          },
        ],
      }
    }
    onClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <VersionChip
          browser="and_chr-69"
          defaultChecked={false}
          isIncluded={false}
          onChange={[Function]}
          version={69}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "browser": "and_chr-69",
          "defaultChecked": false,
          "isIncluded": false,
          "onChange": [Function],
          "version": 69,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <VersionChip
            browser="and_chr-69"
            defaultChecked={false}
            isIncluded={false}
            onChange={[Function]}
            version={69}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "browser": "and_chr-69",
            "defaultChecked": false,
            "isIncluded": false,
            "onChange": [Function],
            "version": 69,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
